datafusion/physical-expr/src/aggregate/count_distinct.rs

```diff
@@ -216,23 +216,19 @@ impl Accumulator for DistinctCountAccumulator {
     }
 
     fn size(&self) -> usize {
+        // temporarily calculating the size approximately, taking first batch size * number of batches
+        // such approach has some inaccuracy for variable length values, like strings.
```

Comment:
This basically remove proper memory accounting for this operation and for strings will likely be very wrong. I would rather see a proper cached size accounting here.

Reply:
Right, as we agreed in https://github.com/apache/arrow-datafusion/issues/5325#issuecomment-1441767680
We need a fix the benchmark and later @alamb can think how to deal with variable length data.

Reply:
This is clearly a conflict of interests: you wanna fix the benchmark, I wanna have proper memory accounting. We likely can have both on the long run, but due to limited development resources, we cannot have the ideal solution right now. 

I'll leave it to the project managers (e.g. @alamb) to decide what's more pressing.

Reply:
> Instead add some code that checks "if is ScalarType::Int8, UInt8, etc then size = size[0]*vec.len()"

https://github.com/apache/arrow-datafusion/issues/5325#issuecomment-1435672826

are we missing this check in this PR?

Reply:
> are we missing this conditionall check in this PR? so we still have accurate size (slow for now) for variable data and accurate size (fast) for fixed lenth data

This is the middle path I would suggest: keep the slow but accurate accounting for variable length data (aka strings) and add a fast path for fixed length sizes (what is in the benchmark)

I believe the the additional overhead of doing accurate size accounting for string values is a relatively smaller amount of the overall time compared to fixed size types. Making `count distinct` with a large number of string values fast is likely going to take a more sophisticated approach to this query in general.

Reply:
@alamb @crepererum Let me remind you again another possible temp solution - The fastest and most painless fix to fix benchmark is to increase batch number, and .size function will be called less often.

Reply:
The overhead for an accurate answer is still enormous for variable length types and gets worse with increasing table sizes, as the `size` gets slower with more distinct values in the aggregation, and is called for every update (so a `O(n^2)` operation).

Reply:
@alamb @crepererum @jychen7 @Dandandan Amended to get approx_size for primitives only